### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by bcdurden** [harvester/harvester#2745]
**Describe the bug**
The Terraform Provider fails to provision VMs 

**To Reproduce**
1. Drop the Terraform code into your working directory and untar
2. Point your kubeconfig context to your harvester cluster
3. Edit the resources.tf file and set the vlan id to something appropriate on your test network that has DHCP
4. Run `terraform init`
5. Run `terraform apply` and choose yes. Terraform code will pull an Ubuntu cloudimg and create a vlan network as well as a VM.
6. Open the Harvester UI and observe the VM never leasing an IP and Terraform spinning
7. Reboot the VM from the UI, watch the VM acquire its IP lease (assuming you have DHCP on your vlan)

**Expected behavior**
The VM should start and acquire an IP lease without needing a manual reboot. When doing this manually in the UI there are no problems. I suspect it has something to do with cloud-init and how Terraform and Harvester handle it in various formats (Harvester has trouble displaying plaintext-sourced cloud-init correctly)

**Environment**
 - Harvester ISO version: 1.0.3
 - Underlying Infrastructure: Bare Metal HPE EL8000, 4-node and Bare Metal HX90 mediabox.

[terraform_code.tar.gz](https://github.com/harvester/harvester/files/9508363/terraform_code.tar.gz)

**Comment by futuretea** [comment]
The content in cloud-init is indented incorrectly and works fine after modification
```diff
  cloudinit {
    type      = "noCloud"
    
    user_data    = <<EOT
-        #cloud-config
-        package_update: true
-        packages:
-        - qemu-guest-agent
-        runcmd:
-        - - systemctl
-            - enable
-            - '--now'
-            - qemu-guest-agent.service
+      #cloud-config
+      package_update: true
+      packages:
+      - qemu-guest-agent
+      runcmd:
+      - - systemctl
+        - enable
+        - '--now'
+        - qemu-guest-agent.service
      EOT
    network_data = ""
  }
```

**Comment by bcdurden** [comment]
I just tested this and found that it makes no difference. If indentation were the case then cloud-init wouldn't work at all. It works fine after reboot. I am able to use more sophisticated cloud-init configurations just fine, but if I Terraform the VM (as opposed to creating it in the UI), it will require a reboot before it can lease an IP address.

In fact, if there is no cloud-init config provided at all, it will still fail.

I suspect this is due to how Terraform is feeding inline cloudinit directly to the VM configuration as opposed to how the UI does it with a secretRef.

**Comment by futuretea** [comment]
@bcdurden can you provide your new terraform code?

**Comment by bcdurden** [comment]
Attached. Also tried a non-empty network_data field without effect.
[terraform.tar.gz](https://github.com/harvester/harvester/files/9527297/terraform.tar.gz)

**Comment by futuretea** [comment]
The content in cloud-init is still indented incorrectly.
```diff
  cloudinit {
    type      = "noCloud"
    
    user_data    = <<EOT
        #cloud-config
        package_update: true
        packages:
        - qemu-guest-agent
        runcmd:
        - - systemctl
-            - enable
-            - '--now'
-            - qemu-guest-agent.service
+        - enable
+        - '--now'
+        - qemu-guest-agent.service
      EOT
    network_data = ""
  }
```

**Comment by futuretea** [comment]
> If indentation were the case then cloud-init wouldn't work at all

No, just caused cloud-init didn't start qemu-guest-agent
After reboot, qemu-guest-agent will be started.


> In fact, if there is no cloud-init config provided at all, it will still fail.

Sure, we need qemu-guest-agent to report IP


> I suspect this is due to how Terraform is feeding inline cloudinit directly to the VM configuration as opposed to how the UI does it with a secretRef.

UI will support read inline cloudinit  in 1.1.0

but that's another issue https://github.com/harvester/harvester/issues/2584

**Comment by bcdurden** [comment]
> The content in cloud-init is still indented incorrectly.
> 
> ```diff
>   cloudinit {
>     type      = "noCloud"
>     
>     user_data    = <<EOT
>         #cloud-config
>         package_update: true
>         packages:
>         - qemu-guest-agent
>         runcmd:
>         - - systemctl
> -            - enable
> -            - '--now'
> -            - qemu-guest-agent.service
> +        - enable
> +        - '--now'
> +        - qemu-guest-agent.service
>       EOT
>     network_data = ""
>   }
> ```

I missed the second indentation error, I thought you were referring to the whole stringified yaml, not a list within it. Testing now.

**Comment by bcdurden** [comment]
~~It is still not working, same behavior. I modified the cloudinit yaml to look like the example in this link: https://registry.terraform.io/providers/harvester/harvester/latest/docs/resources/virtualmachine~~

~~While I did have the error in my current revision, this issue has been a constant problem for the last 2 months, only now has it presented itself with a customer and not in my homelab.~~

~~I think the example you gave above had one too many subtracted indentations, but I understood what you were getting at. Attaching terraform code again.
[terraform.tar.gz](https://github.com/harvester/harvester/files/9527398/terraform.tar.gz)~~

EDIT: Hold, checking something that got reverted in my vlan config. 

EDIT 2: Ok the vlan id got jacked on the last run-through. I am now successfully leasing an IP!
